### PR TITLE
Fix file watching with file replacements

### LINKED ISSUE #11339: Watching doesn't work for file paths remapped by the provided FS host
### Bug Report or Feature Request (mark with an `x`)
```
- [X] bug report -> please search issues before submitting
- [ ] feature request
```

### Area
```
- [X] devkit - @ngtools/webpack
- [ ] schematics
```

### Versions
node v9.5.0
npm 5.7.1
Linux Fedora 25

### Repro steps
Create a new ng cli project. Build it in production mode. As specified by the `angular.json` file, the requests to the `environment.ts` file are remapped to `environment.prod.ts`. However, if you change the `environment.prod.ts` file, that webpack won't rebuild the project. Also, if you change the originally requested file - `environment.ts`, webpack will detect the change and trigger a build.

### Desired functionality
The app should be rebuild when a remapped file is changed.

### Mention any other details that might be useful
It looks like the original paths (and not the remapped ones) are added as dependencies for the webpack compilation in the `@ngtools/webpack` loader - [here](https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/loader.ts#L77-L87). The same thing happens for the original component templates and styles (html, css, scss files).

### PR DESCRIPTION
**Pull request by clydin**
Fixes #11339

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/utils/add-file-replacements.ts [hunk 1]
```diff
@@ -22,53 +29,64 @@ export class MissingFileReplacementException extends BaseException {
...
-  root: Path,
-  host: virtualFs.AliasHost,
-  fileReplacements: FileReplacement[],
-): Observable<null> {
+export interface NormalizedFileReplacement {
+  replace: Path;
+  with: Path;
+}
 
+export function normalizeFileReplacements(
```
Comment (filipesilva) [inline packages/angular_devkit/build_angular/src/utils/add-file-replacements.ts]: Nit: file name does not match export name.

File: tests/legacy-cli/e2e/tests/build/rebuild-replacements.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { getGlobalVariable } from '../../utils/env';
```
Comment (filipesilva) [inline tests/legacy-cli/e2e/tests/build/rebuild-replacements.ts]: WDYT of replacing this with a new spec large in `replacements_spec_large`?

This is the one I had on my branch:

```
it('file replacements work with watch mode', (done) => {
    const overrides = {
      fileReplacements: [
        {
          replace: normalize('/src/meaning.ts'),
          with: normalize('/src/meaning-too.ts'),
        },
      ],
      watch: true,
    };

    let buildNumber = 0;

    runTargetSpec(host, browserTargetSpec, overrides, 45000, createConsoleLogger()).pipe(
      debounceTime(1000),
      tap((buildEvent) => expect(buildEvent.success).toBe(true)),
      tap(() => {
        const fileName = join(outputPath, 'main.js');
        buildNumber += 1;

        switch (buildNumber) {
          case 1:
            expect(virtualFs.fileBufferToString(host.scopedSync().read(fileName)))
              .toMatch(/meaning\s*=\s*42/);
            expect(virtualFs.fileBufferToString(host.scopedSync().read(fileName)))
              .not.toMatch(/meaning\s*=\s*10/);
            host.writeMultipleFiles({
              'src/meaning-too.ts': 'export var meaning = 84;',
            });
            break;

          case 2:
            expect(virtualFs.fileBufferToString(host.scopedSync().read(fileName)))
              .toMatch(/meaning\s*=\s*84/);
            expect(virtualFs.fileBufferToString(host.scopedSync().read(fileName)))
              .not.toMatch(/meaning\s*=\s*42/);
            break;
        }
      }),
      take(2),
    ).toPromise().then(done, done.fail);
  });
```

It should be faster than the e2e and also test the content is correct.

**Comment by filipesilva** [comment]
Also I think this fixes https://github.com/angular/angular-cli/issues/7305

File: packages/angular_devkit/build_angular/src/utils/add-file-replacements.ts (hunk 1 above)
Comment (clydin) [reply packages/angular_devkit/build_angular/src/utils/add-file-replacements.ts]: Done.

File: tests/legacy-cli/e2e/tests/build/rebuild-replacements.ts (hunk 2 above)
Comment (clydin) [reply tests/legacy-cli/e2e/tests/build/rebuild-replacements.ts]: Added.

**Comment by filipesilva** [comment]
LGTM, great work!

**Comment by hansl** [comment]
I fear this doesn't cover the NativeScript case with pattern matching (matching all `.html.tns` files to `.tns`). They'd have to list all those files, and even then it doesn't work when creating new files...

**Comment by clydin** [comment]
Based on the custom host found [here](https://github.com/NativeScript/nativescript-dev-webpack/blob/master/host/platform.ts), code similar to the following should work:
```
            new AngularCompilerPlugin({
                entryModule: resolve(appPath, "app.module#AppModule"),
                tsConfigPath: join(__dirname, "tsconfig.esm.json"),
                skipCodeGeneration: !aot,
                hostReplacementPaths: path => {
                    const { dir, name, ext } = parse(path);

                    for (const platform of extensions) {
                        const platformFileName = `${name}.${platform}${ext}`;
                        const platformPath = toSystemPath(join(dir, platformFileName));
            
                        try {
                            const stat = statSync(platformPath);
                            if (stat && stat.isFile()) {
                                return platformPath;
                            }
                        } catch(_e) {
                            // continue checking the other platforms
                        }
                    }
            
                    return path;
                },
            }),
```

Another option with this PR is to try to use the [PlatformFSPlugin](https://github.com/NativeScript/nativescript-dev-webpack/blob/master/plugins/PlatformFSPlugin.ts) which appears to be used within the non-Angular demos found in the repository.

**Pull request merged**